scripts/o3de/o3de/compatibility.py

```diff
@@ -0,0 +1,245 @@
+    """
+    Returns any incompatible objects for this gem and project.
+    :param gem_json_data: gem json data dictionary
+    :param project_path: path to the project
+    :param all_gems_json_data: optional dictionary containing data for all gems to use in compatibility checks. 
+    If not provided, uses all gems from the manifest, engine and project.
+    """
+    # early out if this project has no assigned engine
+    engine_path = manifest.get_project_engine_path(project_path=project_path)
+    if not engine_path:
+        return set()
+
+    project_json_data = manifest.get_project_json_data(project_path=project_path)
+    if not project_json_data:
+        logger.error(f'Failed to load project.json data from {project_path} needed for checking compatibility')
+        return set(f'project.json (missing)') 
+
+    # in the future we should check if the gem and project depend on conflicting engines and/or apis
+    # we need some way of doing version specifier overlap checks
+
+    engine_json_data = manifest.get_engine_json_data(engine_path=engine_path)
+    if not engine_json_data:
+        logger.error(f'Failed to load engine.json data based on the engine field in project.json or detect the engine from the current folder')
+        return set(f'engine.json (missing)') 
+
+    if not all_gems_json_data:
+        all_gems_json_data = manifest.get_gems_json_data_by_name(engine_path, project_path, include_manifest_gems=True)
+
+    # compatibility will be based on the engine the project uses and the gems visible to
+    # the engine and project
+    return get_gem_engine_incompatible_objects(gem_json_data, engine_json_data, all_gems_json_data)
+
+
+def get_gem_engine_incompatible_objects(gem_json_data:dict, engine_json_data:dict, all_gems_json_data:dict) -> set:
+    """
+    Returns any incompatible objects for this gem and engine.
+    :param gem_json_data: gem json data dictionary
+    :param engine_json_data: engine json data dictionary
+    :param all_gems_json_data: json data of all gems to use for compatibility checks
+    """
+    incompatible_objects = get_incompatible_objects_for_engine(gem_json_data, engine_json_data)
+    return incompatible_objects.union(get_incompatible_gem_dependencies(gem_json_data, all_gems_json_data))
+
+
+def get_incompatible_objects_for_engine(object_json_data:dict, engine_json_data:dict) -> set:
+    """
+    Returns any incompatible objects for this object and engine.
+    If a compatible_engine entry only has an engine name, it is assumed compatible with every engine version with that name.
+    If an engine_api_version entry only has an api name, it is assumed compatible with every engine api version with that name.
+    :param object_json_data: the json data for the object 
+    :param engine_json_data: the json data for the engine
+    """
+    compatible_engines = object_json_data.get('compatible_engines','')
+    engine_api_version_specifiers = object_json_data.get('engine_api_dependencies','')
```

Comment:
FYI: By default, ```get``` will return None if it can't locate from a key, so no need to set the second argument
```suggestion
    compatible_engines = object_json_data.get('compatible_engines')
    engine_api_version_specifiers = object_json_data.get('engine_api_dependencies')
```